src/transformers/optimization.py

```diff
@@ -940,6 +946,325 @@ def get_lr(self):
         return lrs
 
 
+class AdamMini(Optimizer):
+    """
+    Implements Adam Mini algorithm as introduced in [Adam-mini: Use Fewer Learning Rates To Gain More]
+    (https://arxiv.org/abs/2406.16793).
+
+    Parameters:
+        model (`Union[PreTrainedModel, nn.Module]`):
+            The model being trained.
+        lr (`float`, *optional*, defaults to 0.001):
+            The learning rate to use.
+        betas (`Tuple[float,float]`, *optional*, defaults to `(0.9, 0.999)`):
+            Adam's betas parameters (b1, b2).
+        eps (`float`, *optional*, defaults to 1e-06):
+            Adam's epsilon for numerical stability.
+        weight_decay (`float`, *optional*, defaults to 0.0):
+            Decoupled weight decay to apply.
+        model_sharding (`bool`, *optional*, defaults to `False`):
+            Set to True if you are using model parallelism with more than 1 GPU, including FSDP and
+            zero_1,2,3 in Deepspeed. Set to False if otherwise.
+        n_feature (`int`, *optional*):
+            Dimension for hidden features. Can be left unspecified if training non-transformer models.
+        n_head (`int`, *optional*):
+            Number of attention heads. Can be left unspecified if training non-transformer models.
+        n_kv_head (`int`, *optional*):
+            Number of heads for Key and Value. Or equivalently, number of query groups in Group Query
+            Attention. Also known as "n_query_groups". If not specified, it will be equal to n_head.
+            Can be left unspecified if training non-transformer models.
+    """
+
+    def __init__(
+        self,
+        model: Union[PreTrainedModel, nn.Module],
```

Comment:
Why does this optimizer take `model` here instead of `params` like the other optimizers?

Reply:
Adam Mini uses a single momentum and variance for groups of parameters, rather than for each one individually like Adam does. These groups are usually based on layers or modules. To make this work, Adam Mini takes the model and creates these shared parameters for each layer or module instead of doing it for every single parameter.

Reply:
I think to be able to use this with the library the class will need to have the same API as other optimizers. Could you run a test with trainer to see if it integrates properly:

```py
config = LlamaConfig(
    vocab_size=100, 
    hidden_size=32, 
    num_hidden_layers=3, 
    num_attention_heads=4
)
tiny_llama = LlamaForCausalLM(config)
x = torch.randint(0, 100, (128,))
train_dataset = RepeatDataset(x)

with tempfile.TemporaryDirectory() as tmpdir:
    # Trainer without inf/nan filter
    args = TrainingArguments(
        tmpdir,
        learning_rate=1e-9,
        logging_steps=5,
        optim="adam_mini",
    )
    trainer = Trainer(tiny_llama, args, train_dataset=train_dataset)

    # Check this works
    _ = trainer.train()
```

Reply:
> I think to be able to use this with the library the class will need to have the same API as other optimizers. Could you run a test with trainer to see if it integrates properly:
> 
> ```python
> config = LlamaConfig(
>     vocab_size=100, 
>     hidden_size=32, 
>     num_hidden_layers=3, 
>     num_attention_heads=4
> )
> tiny_llama = LlamaForCausalLM(config)
> x = torch.randint(0, 100, (128,))
> train_dataset = RepeatDataset(x)
> 
> with tempfile.TemporaryDirectory() as tmpdir:
>     # Trainer without inf/nan filter
>     args = TrainingArguments(
>         tmpdir,
>         learning_rate=1e-9,
>         logging_steps=5,
>         optim="adam_mini",
>     )
>     trainer = Trainer(tiny_llama, args, train_dataset=train_dataset)
> 
>     # Check this works
>     _ = trainer.train()
> ```

I think they at least need to get a `named_parameters`, like
```
from adam_mini import Adam_mini

optimizer = Adam_mini(
            named_parameters = model.named_parameters(), 
            lr = lr, 
            betas = (beta1,beta2), 
            eps = eps,
            weight_decay = weight_decay, 
            model_sharding = True,
            dim = model_config.dim,
            n_heads = model_config.n_heads,
            n_kv_heads = model_config.n_kv_heads,
            )
```
I think it's impossible to get what they need from the parameters passed to the optimizer. And as long as there is not an optimizer augments, any type of workarounds also seems unfeasible.

Reply:
I’m sorry, but I’m having trouble understanding your point. There are other optimizers, like Lomo and AdaLomo, that also use a model as input instead of parameters. I’m unsure if the issue is due to the different API formats of the optimizers in optimization.py, which might cause Adam Mini to be inconsistent, or if the real problem is that we shouldn’t be passing the model as a parameter to the optimizer. Could you clarify this for me?

Reply:
> I’m sorry, but I’m having trouble understanding your point. There are other optimizers, like Lomo and AdaLomo, that also use a model as input instead of parameters. I’m unsure if the issue is due to the different API formats of the optimizers in optimization.py, which might cause Adam Mini to be inconsistent, or if the real problem is that we shouldn’t be passing the model as a parameter to the optimizer. Could you clarify this for me?

I guess her point is that all current optimizers in transformers lib pass model.parameters as input, and she want Adam mini to stick with that.

Reply:
@Mrw33554432 Right, or at least there should be a test which confirms this optimizer can be selected and run correctly in `Trainer` with `optim="adam_mini"` set

Reply:
@amyeroberts  Instead of passing the model as an argument, can I pass the named parameters as @Mrw33554432 suggested? If that's not possible, I can go ahead and add the test case with Trainer as you proposed.

Reply:
I'd actually like you to do both: pass in the parameters and check that the optimizer can be used in trainer as expected. For the second case, you don't have to add a test, but we should confirm it works. The test that's added here should be updated to have the same structure as the other tests for custom optimizers i.e. the same structure as `test_adafactor`

Reply:
I've written the test case for the trainer, and it's working as expected. However, I can't pass named parameters directly to the optimizer and instead have to pass the entire model. Is there any other workaround for this?